File: tests/test_platform_web_runref.py
"""Guards on clickable run references in chat (issue #241).

The parse and the dispatch registry are *executed* by
``tests/test_platform_web_runref.js``; what is read here is the wiring a DOM-less
test cannot execute, and the two constraints the issue set: the dispatch is
in-app (so the levers that would make a run view addressable are read for by
name), and a reference can only select a view.
"""

import subprocess
from pathlib import Path

from tests.conftest import node_binary, require_node_toolchain

WEB = Path(__file__).resolve().parent.parent / "web"
RUNREF = WEB / "src" / "runref.js"
APP = WEB / "src" / "App.vue"
MARKDOWN = WEB / "src" / "components" / "Markdown.vue"
TASKDEF = (
    Path(__file__).resolve().parent.parent / "taskdef" / "orchestrate" / "instructions.txt"
)


def _read(path):
    return path.read_text(encoding="utf-8")


def test_the_parser_and_the_registry_behave():
    """Execute the module — the parse is this feature's security boundary."""
    node = node_binary()
    if not node:
        require_node_toolchain("no Node available (run `lmer platform setup-ui`)")

    script = Path(__file__).resolve().parent / "test_platform_web_runref.js"
    result = subprocess.run(
        [node, str(script)], capture_output=True, text=True, timeout=60,
    )
    assert result.returncode == 0, result.stderr or result.stdout
    assert "all assertions passed" in result.stdout


def test_the_renderer_reuses_the_module_s_shape():
    """A second copy is how a reference renders as a link the dispatcher then
    refuses to parse — a dead link with nothing on screen to say why."""
    text = _read(MARKDOWN)
    assert "from '../runref.js'" in text
    assert "RUN_REF_RE" in text
    assert "const RUN_REF_RE" not in text, (
        "Markdown.vue defines its own reference pattern — import it from runref.js"
    )
    # A regexp literal here would be that second copy under another name.
    assert "lmer:\\/\\/" not in text, (
        "Markdown.vue builds its own lmer: pattern — import RUN_REF_RE instead"
    )


def test_both_click_buttons_are_intercepted():
    """Both events, one handler — and only the middle button.

    `click` is the primary button only, so a middle click bypassed the handler
    entirely; `auxclick` is every non-primary button, so binding it without a
    filter made a right-click dispatch as well.
    """
    text = _read(MARKDOWN)
    assert '@click="onClick"' in text
    assert '@auxclick="onClick"' in text, (
        "a middle click bypasses a click-only handler and reaches the browser"
    )
    # The binding was right while the behaviour was wrong, which is what the
    # browser rig is for: this guard can see the string, not the dispatch.
    assert "event.button !== 1" in text, (
        "auxclick is bound without a button filter, so a right-click dispatches"
    )


def test_the_dispatch_is_not_navigation():
    """No router, no address bar. Read by name, because "we did not do that" is
    what a later convenience patch removes without noticing."""
    for path in (RUNREF, MARKDOWN):
        text = _read(path)
        for lever in ("window.location", "location.href", "history.pushState",
                      "useRouter", "router.push"):
            assert lever not in text, f"{path.name} navigates via {lever}"


def test_the_reference_carries_no_action():
    """Three key segments and the key: a field that could name a verb, a URL or a
    payload is what turns a view selector into a remote control."""
    lines = [
        line.strip() for line in _read(RUNREF).splitlines()
        if line.strip().startswith("return { host")
    ]
    assert lines == [
        "return { host, project, slug, key: `${host}/${project}/${slug}` }"
    ], lines


def test_the_shell_registers_a_resolver_and_reports_an_untracked_run():
    """App.vue owns tracked-vs-untracked because App.vue holds the fleet: the
    registration, and the notice that keeps an untracked key from being a dead
    link."""
    text = _read(APP)
    assert "onRunRef" in text and "stopRunRefs" in text, (
        "App.vue must register a run-reference resolver and drop it on unmount"
    )
    assert "runRefNotice" in text
    assert "does not track that run" in text
    # Against the fleet this shell already holds, not by a fetch of its own.
    resolver = text[text.index("function openRunRef"):]
    resolver = resolver[:resolver.index("\n}")]
    assert "runs.value.find" in resolver
    assert "open(run)" in resolver


def test_the_assistant_is_told_to_write_them():
    """A carrier nobody writes references into changes nothing (#241's "done
    when")."""
    text = _read(TASKDEF)
    assert "lmer://run/<host>/<project>/<slug>" in text
    # The two properties the note has to carry.
    assert "select a view" in text
    assert "Slack" in text

File: tests/test_platform_web_runref.js
// Executable checks for run references — the parser and the dispatch registry
// (issue #241). Run by tests/test_platform_web_runref.py.
//
// The parse is the security boundary: anything that is not exactly a reference has
// to come out `null`, because a too-loose pattern means a live link the renderer
// built and the dispatcher then acts on.

import assert from 'node:assert/strict'
import {
  RUN_REF_RE, dispatchRunRef, onRunRef, parseRunRef,
} from '../web/src/runref.js'

// --- what a reference is ----------------------------------------------------
const ref = parseRunRef('lmer://run/gitlab.example.com/acme/widget/develop-issue-381')
assert.deepEqual(ref, {
  host: 'gitlab.example.com',
  project: 'acme/widget',
  slug: 'develop-issue-381',
  key: 'gitlab.example.com/acme/widget/develop-issue-381',
})

// A project path can be deeper than one segment; the slug is always the last.
assert.equal(
  parseRunRef('lmer://run/gitlab.example.com/gh/acme/widget-docs/review-mr-9').project,
  'gh/acme/widget-docs',
)

// Surrounding whitespace is what a markdown link can carry; the key is not.
assert.equal(
  parseRunRef('  lmer://run/h/p/s  ').key,
  'h/p/s',
)

// --- what a reference is not -------------------------------------------------
const rejected = [
  // Not enough to name a run.
  'lmer://run/host/slug',
  'lmer://run/host',
  'lmer://run/',
  // Anything riding along — a reference may only select a view.
  'lmer://run/h/p/s?spawn=1',
  'lmer://run/h/p/s#x',
  'lmer://run/h/p/s/../../etc',
  'lmer://spawn/h/p/s',
  'lmer://run/h/p/s%20',
  'lmer://run/h/p/s https://evil.example',
  // Other schemes, including ones that look like this one.
  'lmerx://run/h/p/s',
  'javascript:alert(1)',
  'https://example.com/run/h/p/s',
  'data:text/html,<script>alert(1)</script>',
  // Empty segments and separators the key format would not survive.
  'lmer://run/h//s',
  'lmer://run/h/p/',
  'lmer://run/-h/p/s',
  // Not a string at all.
  null,
  undefined,
  42,
  {},
]
for (const bad of rejected) {
  assert.equal(parseRunRef(bad), null, `parsed a non-reference: ${String(bad)}`)
  if (typeof bad === 'string') {
    assert.equal(RUN_REF_RE.test(bad.trim()), false, `regexp admits ${bad}`)
  }
}

// --- the dispatch registry ---------------------------------------------------
// Nothing listening is a real state: the renderer is imported by surfaces with no
// fleet to select out of.
assert.equal(dispatchRunRef(ref), false)

let seen = null
const stop = onRunRef((got) => { seen = got })
assert.equal(dispatchRunRef(ref), true)
assert.equal(seen.key, ref.key)

// A null ref must not reach the resolver.
seen = null
assert.equal(dispatchRunRef(null), false)
assert.equal(seen, null)

// Teardown belongs to whoever registered: after it, taps are inert again.
stop()
seen = null
assert.equal(dispatchRunRef(ref), false)
assert.equal(seen, null)

// A stale teardown must not unregister somebody else's resolver.
const first = onRunRef(() => { seen = 'first' })
const second = onRunRef(() => { seen = 'second' })
first()
assert.equal(dispatchRunRef(ref), true)
assert.equal(seen, 'second')
second()

console.log('all assertions passed')
